File: camstudy_app/lib/services/auth_service.dart
import 'dart:async';

import 'package:firebase_auth/firebase_auth.dart';

class AuthService {
  AuthService._();
  static final AuthService shared = AuthService._();

  // A getter (not a field initializer) so constructing AuthService — which happens
  // eagerly via the `shared` singleton — never touches Firebase before it's initialized.
  FirebaseAuth get _auth => FirebaseAuth.instance;

  User? get currentUser => _auth.currentUser;

  Future<User> signUp({required String email, required String password}) async {
    final result = await _auth.createUserWithEmailAndPassword(
      email: email,
      password: password,
    );
    final user = result.user;
    if (user == null) throw Exception('알 수 없는 오류가 발생했습니다.');
    return user;
  }

  Future<User> signIn({required String email, required String password}) async {
    final result = await _auth.signInWithEmailAndPassword(
      email: email,
      password: password,
    );
    final user = result.user;
    if (user == null) throw Exception('알 수 없는 오류가 발생했습니다.');
    return user;
  }

  Future<void> signOut() => _auth.signOut();

  /// Sends an SMS verification code to [phoneNumber] (E.164 format, e.g. "+821012345678").
  /// Only supported on Android/iOS/Web — see PlatformSupport.isPhoneAuthSupported.
  Future<String> sendPhoneVerificationCode(String phoneNumber) {
    final completer = Completer<String>();
    _auth.verifyPhoneNumber(
      phoneNumber: phoneNumber,
      verificationCompleted: (_) {},
      verificationFailed: (FirebaseAuthException error) {
        if (!completer.isCompleted) completer.completeError(error);
      },
      codeSent: (String verificationId, int? resendToken) {
        if (!completer.isCompleted) completer.complete(verificationId);
      },
      codeAutoRetrievalTimeout: (_) {},
    );
    return completer.future;
  }

  /// Verifies the SMS code. When [linkToCurrentUser] is true, the phone credential is
  /// linked to the already-signed-in (email/password) account instead of starting a new session.
  Future<User> verifyPhoneCode({
    required String verificationId,
    required String smsCode,
    required bool linkToCurrentUser,
  }) async {
    final credential = PhoneAuthProvider.credential(
      verificationId: verificationId,
      smsCode: smsCode,
    );

    if (linkToCurrentUser && currentUser != null) {
      final result = await currentUser!.linkWithCredential(credential);
      final user = result.user;
      if (user == null) throw Exception('알 수 없는 오류가 발생했습니다.');
      return user;
    }

    final result = await _auth.signInWithCredential(credential);
    final user = result.user;
    if (user == null) throw Exception('알 수 없는 오류가 발생했습니다.');
    return user;
  }
}

File: CamStudy/CamStudy/Services/AuthService.swift
import FirebaseAuth

enum AuthServiceError: LocalizedError {
    case unknown

    var errorDescription: String? {
        switch self {
        case .unknown:
            return "알 수 없는 오류가 발생했습니다."
        }
    }
}

final class AuthService {
    static let shared = AuthService()

    private init() {}

    var currentUser: User? {
        Auth.auth().currentUser
    }

    func signUp(email: String, password: String, completion: @escaping (Result<User, Error>) -> Void) {
        Auth.auth().createUser(withEmail: email, password: password) { result, error in
            if let error {
                completion(.failure(error))
                return
            }
            guard let user = result?.user else {
                completion(.failure(AuthServiceError.unknown))
                return
            }
            completion(.success(user))
        }
    }

    func signIn(email: String, password: String, completion: @escaping (Result<User, Error>) -> Void) {
        Auth.auth().signIn(withEmail: email, password: password) { result, error in
            if let error {
                completion(.failure(error))
                return
            }
            guard let user = result?.user else {
                completion(.failure(AuthServiceError.unknown))
                return
            }
            completion(.success(user))
        }
    }

    func signOut() throws {
        try Auth.auth().signOut()
    }

    /// Sends an SMS verification code to `phoneNumber` (E.164 format, e.g. "+821012345678").
    func sendPhoneVerificationCode(phoneNumber: String, completion: @escaping (Result<String, Error>) -> Void) {
        PhoneAuthProvider.provider().verifyPhoneNumber(phoneNumber, uiDelegate: nil) { verificationID, error in
            if let error {
                completion(.failure(error))
                return
            }
            guard let verificationID else {
                completion(.failure(AuthServiceError.unknown))
                return
            }
            completion(.success(verificationID))
        }
    }

    /// Verifies the SMS code. When `linkToCurrentUser` is true, the phone credential is
    /// linked to the already-signed-in (email/password) account instead of starting a new session.
    func verifyPhoneCode(
        verificationID: String,
        code: String,
        linkToCurrentUser: Bool,
        completion: @escaping (Result<User, Error>) -> Void
    ) {
        let credential = PhoneAuthProvider.provider().credential(
            withVerificationID: verificationID,
            verificationCode: code
        )

        if linkToCurrentUser, let currentUser = Auth.auth().currentUser {
            currentUser.link(with: credential) { result, error in
                if let error {
                    completion(.failure(error))
                    return
                }
                guard let user = result?.user else {
                    completion(.failure(AuthServiceError.unknown))
                    return
                }
                completion(.success(user))
            }
        } else {
            Auth.auth().signIn(with: credential) { result, error in
                if let error {
                    completion(.failure(error))
                    return
                }
                guard let user = result?.user else {
                    completion(.failure(AuthServiceError.unknown))
                    return
                }
                completion(.success(user))
            }
        }
    }
}
